zeromq: add port getter for streaming zmq blocks

n-west opened the pull request:
Add a port getter/accessor that parses the zmq endpoint. If it is
tcp, then parse out the port and return as an integer. New QA tests
- the port accessor with a known port
- the loopback test for pubsub that already exists, but uses
  an ephermeral port for the sink that is read back for the source
  to attach to

---

Current issue that I have with this is that it requires some string parsing and that the message blocks don't inherit from base_{source,sink}_impl so they don't get the feature. Also, we might want to make this more generic as an endpoint() accessor and force users of it to parse the port on their own. That's a design choice I don't have strong feelings on since I only care about the port and this makes it easier for me although it is less generic and potentially more error prone due to string parsing.

smunaut commented on gr-zeromq/lib/base_impl.cc:
```diff
@@ -54,8 +54,29 @@ namespace gr {
+
 
-    base_sink_impl::base_sink_impl(int type, size_t itemsize, size_t vlen, char *address, int timeout, bool pass_tags, int hwm)
+      base_sink_impl::base_sink_impl(int type, size_t itemsize, size_t vlen, char *address, int timeout, bool pass_tags, int hwm)
```
Wrong white space change

smunaut commented on gr-zeromq/lib/base_impl.h:
```diff
@@ -35,6 +35,8 @@ namespace gr {
     public:
       base_impl(int type, size_t itemsize, size_t vlen, int timeout, bool pass_tags);
       virtual ~base_impl();
+        int port();
```
Wrong indent

smunaut commented on gr-zeromq/include/gnuradio/zeromq/pub_sink.h:
```diff
@@ -57,6 +57,11 @@ namespace gr {
+      /*!
+       * \brief If the endpoint is TCP return the port number; -1 otherwise.
+       */
+      virtual int port() = 0;
```
Instead of adding this to all the variations, the best would be to have a base.h for all the ZMQ sink/source and put the common methods in there.

n-west replied:
indeed; do you know of a reason the message variants don't inherit from the base_impl? thanks for catching the whitespace

smunaut replied:
Currently most of the common implementation code wouldn't have been applicable to the message ones basically ... so it didn't make sense to use the same base class for stream vs message.

And making it a 3 level   base_impl <- base_{stream,message}_impl <- {pub,sub,..}_stream_impl    would have been quite useless because base_impl ended up being empty of any code.

So I opted to just do a separate base class for the stream and left the message ones for someone else to refactor.

jmcorgan commented on gr-zeromq/python/zeromq/qa_zeromq_pubsub.py:
```diff
@@ -48,5 +48,28 @@ def test_001 (self):
+    def test_002(self):
+        # Test getting the port returns a port we explicitly set endpoint to
+        zeromq_pub_sink = zeromq.pub_sink(gr.sizeof_float, 1, "tcp://127.0.0.1:5556", 0)
+        self.assertEqual(5556, zeromq_pub_sink.port())
```
One issue that might arise from hard coding the port number here is if there are other tests that do the same, and run in parallel. Not sure I know the answer yet but I wanted to bring this up.

n-west replied:
Of course and that's likely an issue with existing unit tests. One option is to change all of the zmq unit tests to use ephemeral ports. Do we have anything to gain/lose by never hard coding a port? Probably nothing to lose by replacing the existing test_001 with the new test_003

jmcorgan replied:
I hadn't seen the hard-coding in the other QA tests and agree we shouldn't be doing that. If you want to make that change, make it a new PR though after this one is dealt with.

n-west commented:
Changed to be the more generic endpoint and added a commit to update QA tests. I don't think having a public base class will work because of the required virtual inheritance... or at least I couldn't get it to work.

jmcorgan commented:
I may take a stab at refactoring this to use inheritance the way @smunaut suggested.  It's more complex than it appears as you have to have a base API class for the common user exposed methods, and a separate base impl class with the common data members/methods.  See how gr-uhd does this by splitting out usrp_block.h for the common API and having a usrp_block_impl in lib/ directory.

smunaut commented:
If you want a working example of the inheritance model, gr-fosphor does it :p  Took me some time to work out all the C++ kinks and virtual inheritance and constructor and all that stuff.

jmcorgan commented:
Really, really happy to see progress on this.

n-west commented:
:+1:

n-west commented:
The only problem now is API divergence between streaming and msg blocks

jmcorgan commented:
It's too bad the make functions can't be hoisted into a common one in stream_base, as they are all identical.

jmcorgan commented on gr-zeromq/lib/base_impl.cc:
```diff
@@ -31,8 +31,8 @@
 
-    base_impl::base_impl(int type, size_t itemsize, size_t vlen, int timeout, bool pass_tags)
-      : d_vsize(itemsize * vlen), d_timeout(timeout), d_pass_tags(pass_tags)
+    base_impl::base_impl(size_t type, size_t itemsize, size_t vlen, int timeout, bool pass_tags)
```
The 'type' argument is of type 'int' in the zmq::socket_t constructor, and the current code reflects that, but you've changed it to size_t everywhere.  Why is that?

n-west replied:
I only changed it in this one place because the public API requests a size_t

n-west replied:
I thought it was inconsistently declared between these classes, but now it looks like maybe I was wrong and was looking at the wrong variable.

n-west commented:
They are not identical. They all return a different sptr.

jmcorgan commented:
But they can return the base type sptr--note that in the implementation of the make functions they return a API class sptr wrapped around a derived class instantiation raw pointer.  So, I *think* you could move make() into the stream_base class, have it define a stream_base sptr for its return, then in the _impl.cc just return one of those wrapping the derived class raw pointer.

jmcorgan commented on gr-zeromq/lib/base_impl.cc:
```diff
@@ -54,22 +54,52 @@ namespace gr {
+    {
+      size_t len;
+      char endpoint[1024];
+      d_socket->getsockopt(ZMQ_LAST_ENDPOINT, endpoint, &len);
```
Going into getsockopt() you have to have len set to the size of the endpoint buffer, then it modifies len on the wait out with the actual size of the data it fills it with.  So you need to have len = 1024 and define endpoint with the variable.  And I think this is done differently in gcc vs MSVC.

n-west commented:
Refactoring make functions is left as an exercise for the reader

jmcorgan commented:
@n-west Let me know when you plan to pick this back up.

n-west commented:
new qa for msg blocks passes on first shot, must be right :+1:

jmcorgan commented:
Still need to address the getsockopt issue mentioned above.

n-west commented:
Hate to do this now that it's approved.... but I did leave work today about 1/4 of the way in to refactoring the message blocks. I can either base that off of a new master or we can keep this around a bit longer. I'd kind of like other features (#1080) to be on hold or at least have in mind the refactoring. I really don't want to see API and feature sets diverge between message and stream blocks. That said, I don't know that #1080 actually makes sense with the message based blocks so maybe it's OK.

jmcorgan commented:
No, I was only approving the review I had done earlier; still trying to get the hang of this new github feature.  It wasn't an approval of the whole PR.

Also, I don't think we should be using malloc to deal with the getsockopt buffer.  The way to do this in GCC is to simply allocate the buffer as a local array but use the variable as the size.  This is not portable to MSVC, however, and another way to do it is to use alloca() in place of malloc (and don't free() it afterwards).

Also, happy to see you're working on the message ones, carry on :+1:

n-west commented:
a variable length array would be great and I agree malloc is overkill for this. But like you pointed out MSVC makes this harder and uglier than it needs to be.

jmcorgan commented:
We've dealt with this elsewhere in the code and I think we have a macro for it.

jmcorgan commented:
Ah, yes:

https://github.com/gnuradio/gnuradio/blob/master/gnuradio-runtime/include/gnuradio/attributes.h#L84

n-west commented:
:+1:

n-west commented:
Actually, I need to test the set_endpoint; I'm concerned about what might happen with the rep/req calls.

n-west commented:
repreq test sometimes segfaults and I can't figure out why, but otherwise this seems OK

jmcorgan commented:
Other than that, Mrs. Lincoln, how was the play?

n-west commented:
Phenomenal.

n-west commented:
Doneso. We could be more aggressive about grouping similar code (ex. another base class at API level as a base for *all* zmq blocks) or not (ex. breaking out the readloop/start/stop stuff in msg_base_impl.cc rather than the conditionals in there), but that gets pretty subjective quickly and I'd rather leave it to someone with stronger opinions on the matter.

jmcorgan commented:
Yeah, I think this is a great step forward in modularizing things to allow adding common API methods.  I'll be putting it through the wringer here.

jmcorgan commented:
So, it turns out the ZMQ_LAST_ENDOINT option to getsockopt was only introduced in ZMQ API >=3, and Ubuntu 12.04 only has ZMQ API 2. As 12.04 is still on the list of supported OSes for 3.7, there are two options:

- Conditionalize the endpoint related API functions though #ifdef's so they are only defined if ZMQ_LAST_ENDPOINT is defined, or
- Make this refactoring next branch/3.8 only, which drops support for Ubuntu 12.04

I'm going to continue testing the refactoring on 14.04 that has ZMQ API 4.

jmcorgan commented:
With 14.04 this branch consistently fails two QA tests due to segfaults:

----------------------------------------
The following tests passed:
	qa_zeromq_msg_reqrep
	qa_zeromq_pub
	qa_zeromq_pubsub
	qa_zeromq_pushpull
	qa_zeromq_reqrep

71% tests passed, 2 tests failed out of 7

Total Test time (real) =   3.60 sec

The following tests FAILED:
	211 - qa_zeromq_msg_pubsub (Failed)
	212 - qa_zeromq_msg_pushpull (Failed)
Errors while running CTest

n-west commented:
Does that mean it's not segfaulting?

Given the API issue I think moving it to next is OK. Conditionalizing it will only be ugly and being on next-only would certainly move attention that way.

jmcorgan commented:
No, the segfaults happen with 14.04, master, and zmq v3.  I haven't yet tried this on next branch with zmq v4.

I agree with your reasoning regarding master/next.  Github now lets you change the branch to merge into for a PR; you can do this, then rebase your series of commits off current next branch.

I'll test it again at that point, but I don't expect the segfault issues to change.

I was in the middle of another line-by-line review yesterday when a bunch of other stuff came up.  I'll get back to that soon.

n-west commented:
:-/ I'll have to spin up a 14.04 vm to verify. Definitely not segfaulting on my 16.04 machine (zmq 5) and I don't think it was failing on my Debian machine either.

jmcorgan commented:
This is the biggest reason we're moving forward with our own CI build farm, so that every PR will automatically be tested against all our supported platform configurations.  Had this been in place, we'd have learned about the 12.04 API problem weeks ago, and the segfaults soon after you pushed the new branch.

jmcorgan commented:
This needs rebasing off current next branch. It would be great to the results of this with the new CI tester.

jmcorgan commented:
@awalls-cx18 Could you please take a look at this?

jmcorgan commented:
I think a clean way of addressing this would be to squash this to a single commit, cherry pick that commit onto the current next branch, force pushing it to your github repo, then editing the pull request to set the base branch to next.

jmcorgan commented:
I have a branch where I've done this in my repo [here](https://github.com/jmcorgan/gnuradio/commit/ca6504e63b40bb013335b45886d45235a5e250b8). It still is having problems with segfaults and hangs during the ZMQ QA tests.

jmcorgan commented:
See https://ci.gnuradio.org/buildbot/#/builders/10/builds/16 for the failures on the new CI system with this PR redone in #1287. There is a segfault in qa_zeromq_msg_pubsub, a test failure in qa_zeromq_pubsub, and gets hung in qa_zeromq_pub.

jmcorgan commented:
Closing for inactivity.
